File: client/src/domain/combat.ts
import { statusDefinitions } from './stats/statusCatalog';
import { applyStatus, completeStatusActivation } from './stats/statuses';
import { enemyStats } from './stats/resolve';
import { affordability } from './stats/resources';
import type { Immutable } from 'immer';
import type { ActionSnapshot, Character, Enemy } from './model';
import { wikiValue } from './skills/wiki';
import { skills, ranks } from './Skills';
import { createRng } from './rng';
import { emit } from './battle/events';
import {
    defenses,
    snapshotAction,
    train,
    requirementReason,
    effectiveStats,
    equipment,
    refreshStats,
} from './skillSystem';

export function damageAmount(
    power: number,
    multiplier: number,
    defense: number,
    protection: number,
    criticalBonus = 0,
) {
    const mitigated = Math.floor(Math.max(0, power - defense) * multiplier);
    return Math.floor(
        Math.floor(mitigated * (1 + criticalBonus)) * (1 - Math.max(0, Math.min(1, protection))),
    );
}
export function actionPower(a: Immutable<ActionSnapshot>) {
    return (
        a.rank.base +
        (a.attack * (a.rank.attackMultiplier ?? 1)) /
            (a.skill === 'arrowRevolver' ? 1 : skills[a.skill].hits)
    );
}
export function previewDamage(
    c: Immutable<Character>,
    enemy: Immutable<Enemy>,
    id: string,
    critical = false,
): number {
    const a = snapshotAction(c, id, enemy.id, false);
    if (skills[id].effect !== 'attack') return 0;
    const t = a.targets.find((t) => t.id === enemy.id);
    if (!t) return 0;
    const damage =
        damageAmount(actionPower(a), 1, t.defense, t.protection, critical ? a.criticalBonus : 0) *
        skills[id].hits;
    return Math.min(enemy.hp, Math.max(0, damage - (enemy.shield ?? 0)));
}
export function hitEnemy(enemy: Enemy, amount: number) {
    const absorbed = Math.min(enemy.shield ?? 0, amount);
    enemy.shield = (enemy.shield ?? 0) - absorbed;
    const before = enemy.hp;
    enemy.hp = Math.max(0, enemy.hp - (amount - absorbed));
    return before - enemy.hp;
}
export function trainOffense(
    c: Character,
    a: Immutable<Pick<ActionSnapshot, 'skill' | 'melee' | 'sword' | 'dual'>>,
    hit: boolean,
    kills: number,
    criticalHit = false,
    criticalKills = 0,
) {
    for (const id of [
        a.skill,
        ...(a.melee || a.skill === 'normal' ? ['combatMastery'] : []),
        ...(a.sword ? ['swordMastery'] : []),
        ...(a.dual ? ['dualMastery'] : []),
        ...((skills[a.skill].talent ?? equipment(c).weapon?.talent) === 'Magic'
            ? ['magicMastery']
            : []),
        ...((skills[a.skill].talent ?? equipment(c).weapon?.talent) === 'Archery'
            ? ['bowMastery', 'rangeAttack'].filter((id) => id !== a.skill)
            : []),
    ]) {
        if (hit) train(c, id, 'hit');
        if (kills) train(c, id, 'kill', kills);
    }
    if (hit && a.melee && c.effects.final) train(c, 'final', 'buffHit');
    if (criticalHit) train(c, 'critical', 'critical');
    if (criticalKills) train(c, 'critical', 'criticalKill', criticalKills);
}
export function commitAbility(c: Character, a: ActionSnapshot, operationId: string) {
    const b = c.battle!;
    const skill = skills[a.skill];
    const reason = affordability(c, a.costs, false);
    if (reason) throw new Error(reason);
    for (const pool of ['hp', 'mana', 'stamina'] as const)
        c[pool] = Math.round((c[pool] - a.costs[pool]) * 1e10) / 1e10;
    if (skill.effect === 'counter') {
        c.effects.counter = {
            power: actionPower(a),
            opponentMultiplier: a.rank.counterMultiplier ?? 0,
            source: { skill: a.skill, melee: a.melee, sword: a.sword, dual: a.dual },
        };
        train(c, 'counter', 'use');
        b.log.push('Counterattack prepared.');
    } else if (skill.effect === 'buff') {
        c.effects.final = {
            magnitude: a.rank.base,
            remaining: a.rank.duration,
        };
        train(c, 'final', 'use');
        b.log.push(`Final Hit grants +${c.effects.final.magnitude} melee attack.`);
    } else if (skill.effect === 'heal') {
        const amount = Math.floor(a.rank.base * 5);
        const before = c.hp;
        c.hp = Math.min(effectiveStats(c).hp, c.hp + amount);
        train(c, a.skill, 'use');
        b.log.push(`${skill.name} restores ${c.hp - before} HP.`);
    } else if (skill.effect === 'restoreMana') {
        const before = c.mana;
        const max = effectiveStats(c).mana;
        c.mana = Math.min(max, c.mana + Math.floor((max * a.rank.base) / 100));
        train(c, a.skill, 'use');
        b.log.push(`${skill.name} restores ${c.mana - before} MP.`);
    } else if (skill.effect === 'defend') {
        c.statuses = applyStatus(c, 'guard', { id: a.skill, name: skill.name }, true, {
            ...statusDefinitions.guard,
            name: 'Defense',
            modifiers: [
                { stat: 'defense', flat: a.rank.base },
                {
                    stat: 'protection',
                    flat: wikiValue(
                        skill.wiki,
                        'Protection Bonus',
                        ranks.indexOf(a.rank.rank),
                        c.race,
                    ),
                },
            ],
        }).statuses;
        train(c, a.skill, 'use');
        b.log.push('Defense prepared until your next turn.');
    } else if (skill.effect === 'manaShield') {
        c.effects.manaShield = {
            efficiency: wikiValue(
                skill.wiki,
                'Base Mana Efficiency',
                ranks.indexOf(a.rank.rank),
                c.race,
            ),
            upkeep: Math.ceil(
                wikiValue(skill.wiki, 'Mana Use [/sec]', ranks.indexOf(a.rank.rank), c.race),
            ),
            remaining: 3,
        };
        train(c, a.skill, 'use');
        b.log.push('Mana Shield active until your third subsequent turn.');
    } else if (skill.effect === 'status') {
        train(c, a.skill, 'use');
    } else {
        let hit = false,
            kills = 0,
            criticalHit = false,
            criticalKills = 0;
        const rng = createRng(b.rng);
        for (const t of a.targets) {
            const enemy = b.enemies.find((e) => e.id === t.id && e.hp > 0);
            if (!enemy) continue;
            const critical = rng.chance(a.criticalChance / 10000);
            const amount = damageAmount(
                actionPower(a),
                1,
                t.defense,
                t.protection,
                critical ? a.criticalBonus : 0,
            );
            let damage = 0;
            for (let i = 0; i < skill.hits && enemy.hp > 0; i++) {
                const dealt = hitEnemy(enemy, amount);
                damage += dealt;
                emit(c.battle!, operationId, {
                    type: 'damage',
                    actorId: c.id,
                    targetId: enemy.id,
                    amount: dealt,
                    critical,
                    hitIndex: i,
                    text: `${skill.name}${critical ? ' critical' : ''} hits ${enemy.name} for ${dealt}.`,
                });
            }
            hit ||= damage > 0;
            criticalHit ||= critical && damage > 0;
            if (!enemy.hp) {
                kills++;
                if (critical) criticalKills++;
            }
            b.log.push(
                `${skill.name}${critical ? ' critical' : ''} hits ${enemy.name} for ${damage}.`,
            );
        }
        b.rng = rng.snapshot();
        trainOffense(c, a, hit, kills, criticalHit, criticalKills);
        for (const id of [c.equipment.main, ...(a.dual ? [c.equipment.offhand] : [])]) {
            const item = c.inventory.find((i) => i.id === id);
            if (item) item.durability = Math.max(0, (item.durability ?? 20) - 1);
        }
    }
    for (const status of skill.appliedStatuses ?? []) {
        if (status.target === 'self')
            c.statuses = applyStatus(
                c,
                status.id,
                { id: a.skill, name: skill.name },
                true,
            ).statuses;
        else
            for (const target of a.targets) {
                const enemy = b.enemies.find((e) => e.id === target.id && e.hp > 0);
                if (enemy)
                    enemy.statuses = applyStatus(
                        enemy,
                        status.id,
                        { id: a.skill, name: skill.name },
                        false,
                    ).statuses;
            }
    }
    c.cooldowns[a.skill] = a.rank.cooldown;
}
export function endPlayerActivation(c: Character, cast?: string) {
    for (const id of Object.keys(c.cooldowns))
        if (id !== cast) c.cooldowns[id] = Math.max(0, c.cooldowns[id] - 1);
    if (c.effects.final && cast !== 'final' && --c.effects.final.remaining <= 0)
        delete c.effects.final;
    const completed = completeStatusActivation(c);
    for (const pool of ['hp', 'mana', 'stamina'] as const) c[pool] = completed.actor[pool];
    c.statuses = completed.actor.statuses;
    c.stats = completed.actor.stats;
    c.effects = completed.actor.effects;
    c.battle?.log.push(...completed.log);
}
export function enemyHit(
    c: Character,
    e: Enemy,
    power: number,
    status: string | undefined,
    operationId: string,
) {
    const b = c.battle!;
    if (c.effects.counter && (!e.attackType || e.attackType === 'melee')) {
        const stance = c.effects.counter;
        delete c.effects.counter;
        const damage = hitEnemy(
            e,
            damageAmount(
                stance.power + enemyStats(e).attack * (stance.opponentMultiplier ?? 0),
                1,
                enemyStats(e).defense,
                enemyStats(e).protection,
            ),
        );
        train(c, 'counter', 'counter');
        trainOffense(c, stance.source, damage > 0, e.hp === 0 ? 1 : 0);
        emit(b, operationId, {
            type: 'counter',
            actorId: c.id,
            targetId: e.id,
            amount: damage,
            text: `Counterattack negates ${e.name}'s hit and deals ${damage}.`,
        });
        return;
    }
    const d = defenses(c, e.attackType === 'magic');
    const raw = damageAmount(enemyStats(e).attack * power, 1, d.defense, d.protection);
    const absorbed = Math.min(c.effects.shield ?? 0, raw);
    c.effects.shield = (c.effects.shield ?? 0) - absorbed;
    let damage = raw - absorbed;
    const manaShield = c.effects.manaShield;
    if (manaShield) {
        const blocked = Math.min(damage, Math.floor(Math.floor(c.mana) * manaShield.efficiency));
        c.mana -= Math.ceil(blocked / manaShield.efficiency);
        damage -= blocked;
    }
    damage = Math.min(c.hp, damage);
    c.hp -= damage;
    for (const id of ['shieldMastery', 'lightMastery', 'heavyMastery'])
        if (!requirementReason(c, id)) train(c, id, 'incoming');
    if (c.effects.defense) train(c, 'defense', 'use');
    emit(b, operationId, {
        type: 'damage',
        actorId: e.id,
        targetId: c.id,
        amount: damage,
        critical: false,
        hitIndex: 0,
        text: `${e.name} deals ${damage} damage.`,
    });
    if (c.hp > 0 && status)
        c.statuses = applyStatus(c, status, { id: e.id, name: e.name }, false).statuses;
    refreshStats(c);
}

File: game/domain/combat.lua
local C = require("game.content.catalog")
local S = require("game.domain.stats")
local U = require("game.domain.util")
local R = require("game.services.rng")
local Skills = require("game.domain.skills")
local Log = require("game.domain.combat_log")
local M = {}

local function category(p, actor, definition)
	return actor.hero and (definition.category or S.current(p).category) or "melee"
end

local function power(p, actor, damage_category)
	return actor.hero and S.power(p, damage_category) or C.enemies[actor.def].attack
end

function M.status(p, id)
	p.battle.statuses[id] = p.battle.statuses[id] or {}
	return p.battle.statuses[id]
end

local function counters(p, actor, target, definition)
	local effects = p.battle.statuses[target.id] or {}
	return effects.counterattack
		and not definition.bypass_counter
		and definition.hits == 1
		and category(p, actor, definition) == "melee"
end

---@return number chance, number bonus
function M.critical(p, actor, skill)
	if not actor.hero then
		return 0, 0
	end
	local s, d = S.current(p), C.skills[skill]
	local bonus = s.critical_bonus or 0
	if bonus == 0 then
		return 0, 0
	end
	local extra = d.two_handed_critical and s.two_handed and d.two_handed_critical or 0
	return U.clamp((s.critical_chance or 0) + extra, 0, 1), bonus
end

---@return number damage, integer hits, number absorbed, boolean guarded
---Pure pipeline: flat defense -> critical -> protection -> shield. No RNG or writes.
function M.damage(p, actor, target, skill, critical, raw_override, reaction)
	local d = C.skills[skill]
	local damage_category = category(p, actor, d)
	local hits = d.hits or 1
	if not reaction and counters(p, actor, target, d) then
		return 0, hits, 0, false
	end
	local multiplier = actor.hero and p.race == "Giant" and d.giant_multiplier or d.multiplier or 1
	if actor.hero and d.two_handed_multiplier and S.current(p).two_handed then
		multiplier = multiplier * d.two_handed_multiplier
	end
	local raw = raw_override or ((d.base or 0) + power(p, actor, damage_category)) * multiplier
	local defense, protection, absorption = 0, 0, 0
	if target.hero then
		local s = S.current(p)
		defense = damage_category == "magic" and s.magic_defense or s.defense
		protection = damage_category == "magic" and s.magic_protection or s.protection
	else
		defense = C.enemies[target.def].defense
	end
	local effects = p.battle.statuses[target.id] or {}
	local guarded = effects.guard ~= nil and not d.bypass_guard and damage_category ~= "magic" and not reaction
	if guarded then
		defense = defense + effects.guard.defense
		protection = protection + effects.guard.protection
		absorption = target.hero and (S.current(p).shield_absorption or 0) or 0
	end
	if effects.armor_break then
		defense = math.max(0, defense - 4)
	end
	local amount = math.floor(math.max(0, raw / hits - defense))
	local _, bonus = M.critical(p, actor, skill)
	if critical then
		amount = math.floor(amount * (1 + bonus))
	end
	amount = math.floor(amount * (1 - U.clamp(protection, 0, 100) / 100))
	local final = math.floor(amount * (1 - absorption))
	return final, hits, amount - final, guarded
end

---@return number normal, integer hits, number critical, number normal_total, number critical_total
function M.preview(p, actor, target, skill)
	local normal, hits, _, guarded = M.damage(p, actor, target, skill, false)
	local critical = M.damage(p, actor, target, skill, true)
	local normal_total, critical_total = normal * hits, critical * hits
	if guarded and hits > 1 then
		-- Defense is consumed by the first hit. Later hits use the same crit decision.
		normal_total = normal + M.damage(p, actor, target, skill, false, nil, true) * (hits - 1)
		critical_total = critical + M.damage(p, actor, target, skill, true, nil, true) * (hits - 1)
	end
	return normal, hits, critical, normal_total, critical_total
end

local function deal(p, actor, target, skill, damage, absorbed, critical, hit, hits)
	local loss = math.min(target.pools.hp, damage)
	target.pools.hp = target.pools.hp - loss
	Log.emit(p, "damage", target.name .. " takes " .. loss .. " damage" .. (critical and " · Critical" or ""), {
		actor = actor.id,
		target = target.id,
		skill = skill,
		amount = loss,
		actual = loss,
		calculated = damage + absorbed,
		absorbed = absorbed,
		critical = critical,
		hit = hit,
		hits = hits,
	})
	if target.pools.hp <= 0 then
		Log.emit(p, "defeat", target.name .. " is defeated", { target = target.id })
	end
	return loss
end

local function down(p, actor, target)
	M.status(p, target.id).downed = { source = actor.id, applied = p.battle.turn }
	Log.emit(p, "status", target.name .. " is downed until the attacker's next turn ends", {
		actor = actor.id,
		target = target.id,
		status = "downed",
	})
end

---@return number dealt, boolean critical, integer defeats
function M.resolve(p, actor, target, skill, factory)
	local d = C.skills[skill]
	if counters(p, actor, target, d) then
		local effects = M.status(p, target.id)
		local reaction = effects.counterattack
		effects.counterattack = nil
		Log.emit(
			p,
			"status_end",
			target.name .. " counters the attack",
			{ actor = target.id, target = target.id, status = "counterattack" }
		)
		local raw = power(p, actor, "melee") * reaction.enemy_share + power(p, target, "melee") * reaction.self_share
		local damage, _, absorbed = M.damage(p, target, actor, "normal", false, raw, true)
		local loss = deal(p, target, actor, "counterattack", damage, absorbed, false, 1, 1)
		if loss > 0 and actor.pools.hp > 0 then
			down(p, target, actor)
		end
		if target.hero then
			Skills.train(p, {
				id = p.battle.id .. ":" .. p.battle.turn,
				skill = "counterattack",
				damage = loss,
				counter = true,
				counter_negated = true,
				hero_action = false,
			})
		end
		return 0, false, 0
	end
	local chance = M.critical(p, actor, skill)
	-- Exactly one counted number sample per eligible target, including chance 0/1.
	local critical = false
	local _, bonus = M.critical(p, actor, skill)
	if bonus > 0 then
		critical = R.open(p.battle.random, factory):raw() / 4294967296 < chance
	end
	local dealt = 0
	for hit = 1, d.hits do
		if target.pools.hp <= 0 then
			break
		end
		local damage, _, absorbed, blocked = M.damage(p, actor, target, skill, critical)
		dealt = dealt + deal(p, actor, target, skill, damage, absorbed, critical, hit, d.hits)
		if blocked then
			M.status(p, target.id).guard = nil
			Log.emit(p, "status_end", target.name .. " consumes Defense", { target = target.id, status = "guard" })
			if target.hero then
				Skills.train(p, {
					id = p.battle.id .. ":" .. p.battle.turn .. ":guard",
					skill = skill,
					damage = damage,
					hero_action = false,
					guarded = true,
				})
			end
		end
	end
	if target.pools.hp > 0 then
		if d.knockdown and dealt > 0 then
			down(p, actor, target)
		end
		if d.status then
			local remaining = d.status == "poison" and 3 or 2
			M.status(p, target.id)[d.status] = { remaining = remaining, applied = p.battle.turn }
			Log.emit(
				p,
				"status",
				target.name .. " is affected by " .. d.status,
				{ actor = actor.id, target = target.id, status = d.status, remaining = remaining }
			)
		end
	end
	return dealt, critical, target.pools.hp <= 0 and 1 or 0
end
return M
